sdk/iothub/src/service/twin.rs

```diff
@@ -0,0 +1,325 @@
+use azure_core::errors::{extract_status_headers_and_body, AzureError, UnexpectedHTTPResult};
+use hyper::{Body, Client, Method, Request, StatusCode};
+use hyper_tls::HttpsConnector;
+use serde::de::DeserializeOwned;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use crate::service::{ServiceClient, API_VERSION};
+
+/// AuthenticationType of a module or device
+#[derive(Debug, Deserialize)]
+pub enum AuthenticationType {
+    #[serde(rename = "certificate")]
+    Certificate,
+    #[serde(rename = "Authority")]
+    Authority,
+    #[serde(rename = "none")]
+    None,
+    #[serde(rename = "sas")]
+    SAS,
+    #[serde(rename = "selfSigned")]
+    SelfSigned,
+}
+
+/// The connection state of a module or device
+#[derive(Debug, Deserialize)]
+pub enum ConnectionState {
+    Connected,
+    Disconnected,
+}
+
+/// Device or module status
+#[derive(Deserialize, Debug)]
+pub enum Status {
+    #[serde(rename = "disabled")]
+    Disabled,
+    #[serde(rename = "enabled")]
+    Enabled,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct DeviceCapabilities {
+    #[serde(rename = "iotEdge")]
+    pub iotedge: bool,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct X509ThumbPrint {
+    pub primary_thumbprint: Option<String>,
+    pub secondary_thumbprint: Option<String>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct TwinProperties {
+    pub desired: serde_json::Value,
+    pub reported: serde_json::Value,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct DeviceTwin {
+    pub authentication_type: AuthenticationType,
+    pub capabilities: DeviceCapabilities,
+    pub cloud_to_device_message_count: i64,
+    pub connection_state: ConnectionState,
+    pub device_etag: String,
+    pub device_id: String,
+    pub device_scope: Option<String>,
+    pub etag: String,
+    pub last_activity_time: String,
+    pub parent_scopes: Option<Vec<String>>,
+    pub properties: TwinProperties,
+    pub status: Status,
+    pub status_reason: Option<String>,
+    pub status_update_time: String,
+    pub tags: HashMap<String, String>,
+    pub version: i64,
+    pub x509_thumbprint: X509ThumbPrint,
+}
+
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ModuleTwin {
+    pub authentication_type: AuthenticationType,
+    pub cloud_to_device_message_count: i64,
+    pub connection_state: ConnectionState,
+    pub device_etag: String,
+    pub device_id: String,
+    pub etag: String,
+    pub last_activity_time: String,
+    pub module_id: String,
+    pub properties: TwinProperties,
+    pub status: Status,
+    pub status_update_time: String,
+    pub version: i64,
+    pub x509_thumbprint: X509ThumbPrint,
+}
+
+struct DesiredTwin {
+    contents: serde_json::Value,
+}
+
+pub struct DesiredTwinBuilder<'a, R>
+where
+    R: DeserializeOwned,
+{
+    service_client: &'a ServiceClient,
```

Comment:
I'm unsure about taking a client by reference instead of ownership, but it's probably fine in this case.

Reply:
If we move ownership to DesiredTwinBuilder, `ServiceClient` cannot be used anymore after setting the device/module twin as we move `ServiceClient` out of itself in that case. Any subsequent calls to `ServiceClient` won't compile any longer.

```rust
service_client.update_device_twin()  // service_client moves into DesiredTwinBuilder here
              ...
              .execute();            // DesiredTwinBuilder gets destroyed with ServiceClient

let twin = service_client.get_module_twin(); // Won't compile
```